cmd/oras/manifest/fetchConfig.go

```diff
@@ -0,0 +1,176 @@
+`,
+		Args: cobra.ExactArgs(1),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if opts.outputPath == "-" && opts.OutputDescriptor {
+				return errors.New("`--output -` cannot be used with `--descriptor` at the same time")
+			}
+
+			opts.cacheRoot = os.Getenv("ORAS_CACHE")
+			return opts.ReadPassword()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			return fetchConfig(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.outputPath, "output", "o", "", "output file path")
+	cmd.Flags().StringVarP(&opts.mediaType, "media-type", "", "", "media type of the manifest config")
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func fetchConfig(opts fetchConfigOptions) (fetchErr error) {
+	ctx, _ := opts.SetLoggerLevel()
+
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	if repo.Reference.Reference == "" {
+		return oerrors.NewErrInvalidReference(repo.Reference)
+	}
+
+	var src oras.ReadOnlyTarget = repo
+	if opts.cacheRoot != "" {
+		ociStore, err := oci.New(opts.cacheRoot)
+		if err != nil {
+			return err
+		}
+		src = cache.New(repo, ociStore)
+	}
+
+	configDesc, err := fetchConfigDesc(ctx, src, opts.targetRef, opts.mediaType)
+	if err != nil {
+		return err
+	}
+
+	// output config's descriptor if `--descriptor` is used
+	if opts.OutputDescriptor {
+		descBytes, err := json.Marshal(configDesc)
+		if err != nil {
+			return err
+		}
+		err = opts.Output(os.Stdout, descBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	contentBytes, err := content.FetchAll(ctx, src, configDesc)
+	if err != nil {
+		return err
+	}
+
+	// save config into the local file if the output path is provided
+	if opts.outputPath != "" && opts.outputPath != "-" {
+		file, err := os.Create(opts.outputPath)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := file.Close(); fetchErr == nil {
+				fetchErr = err
+			}
+		}()
+
+		if _, err = file.Write(contentBytes); err != nil {
+			return err
+		}
+	}
+
+	// output config
+	if (opts.outputPath == "" && !opts.OutputDescriptor) || opts.outputPath == "-" {
+		err = opts.Output(os.Stdout, contentBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func fetchConfigDesc(ctx context.Context, src oras.ReadOnlyTarget, reference string, configMediaType string) (ocispec.Descriptor, error) {
+	// fetch manifest descriptor
+	manifestDesc, err := oras.Resolve(ctx, src, reference, oras.ResolveOptions{})
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	// fetch config descriptor
+	successors, err := content.Successors(ctx, src, manifestDesc)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	for i, s := range successors {
+		if s.MediaType == configMediaType || (configMediaType == "" && i == 0 && descriptor.IsImageManifest(manifestDesc.MediaType)) {
+			return s, nil
+		}
+	}
+	return ocispec.Descriptor{}, fmt.Errorf("%s does not have a config", reference)
+}
```

Comment:
Since we know it is an image manifest, we can call `oras.FetchBytes()` and then unmarshal it to extract the config descriptor. The simplified operation not only has less network calls but also cleaner.

Reply:
> Since we know it is an image manifest, we can call `oras.FetchBytes()` and then unmarshal it to extract the config descriptor. The simplified operation not only has less network calls but also cleaner.

Are we sure that `manifest fetch-config` is only used for image manifest? I see in `oras pull --config`, users can provided the media type of the config file. If the manifest is not an image manifest, the command will fetch the blob with matched media type.

Reply:
> Are we sure that `manifest fetch-config` is only used for image manifest?

Yes, manifest fetch-config should only run against image manifests.

Reply:
`oras manifest fetch-config` should return error if the manifest is not an image manifest.

Reply:
Resolved as required.